concepts/websockets/article.md

```diff
@@ -0,0 +1,1073 @@
+  - [ws](#ws)
+    - [Minimal code samples](#minimal-code-samples)
+  - [μWebSockets](#%CE%BCwebsockets)
+    - [Minimal code sample](#minimal-code-sample)
+  - [faye-websocket](#faye-websocket)
+    - [Minimal code samples](#minimal-code-samples)
+  - [Socket.io](#socketio)
+    - [Minimal code samples](#minimal-code-samples)
+  - [SocketCluster](#socketcluster)
+    - [Minimal code samples](#minimal-code-samples)
+  - [SockJS](#sockjs)
+    - [Minimal code samples](#minimal-code-samples)
+  - [deepstream server](#deepstream-server)
+- [Scaling beyond a single server](#scaling-beyond-a-single-server)
+- [Realtime messaging platforms-as-a-service](#realtime-messaging-platforms-as-a-service)
+  - [Ably](#ably)
+  - [PubNub](#pubnub)
+  - [Pusher](#pusher)
+  - [deepstreamHub](#deepstreamhub)
+
+## Aren't we reinventing the wheel here?
+
+One of the best - and worst - things about being a developer is the vast quantity of abstractions available to us to get a job done. So many of the problems we need to solve have been encountered by others before us, that often not only is there usually a significant wealth of wisdom already published that we can draw upon, but often there are numerous solutions already implemented, ready to plug directly into whatever we're building.
+
+Existing abstractions are great, but in some cases they can be a crutch, making us complacent and ultimately limiting what we're actually capable of doing. As the old saying goes, "don't reinvent the wheel", and if you just have a job to get done quickly, then existing solutions are great. Plug them in and move on. This isn't always the right approach, however, and in fact it's rarely that simple anyway.
+
+In order to deeply understand something, we almost always have to experience it ourselves, at least to some degree. Human brains are not optimised to store "facts". They're optimised to capture and match on patterns they experience. This is why, when you let go of your aversion to reinventing wheels and learning new things, and decide to have a go at doing something yourself, the depth of your understanding can grow far beyond what you ever expected and empower you to start coming up with ideas and approaches you'd never have otherwise seen.
+
+There are many WebSocket abstractions out there already, and many more higher-level services built on top of them. By understanding the underlying technology yourself though, when you inevitably hit a brick wall as a result of some mismatch between what you're trying to do and what your chosen abstraction offers, you'll have a way of breaking through that wall. Maybe you'll want to contribute fixes and improvements to whatever you're using, or perhaps you'll choose to go deeper and explore entirely different approaches to your problem - approaches that are highly dependant on knowing what's actually going on under the hood.
+
+As I like to parrot whenever I'm given a soapbox:
```

Comment:
With acknowledgement that you don't need a proper response yet, just a point of consideration that I was trying to put a bit of myself and my personality into it. It's very easy to make for dry reading when going for utilitarian, state-the-facts-only reporting. I'm not overly attached to that intro of course, but personally I think it's interesting, and I thought it might be good to have the article equivalent of a "who is this for" section that you often find in textbooks and other non-fiction books.